<a href="https://colab.research.google.com/github/yasminbv/DataScience/blob/main/Homework5.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Case Intro
Term deposits are a major source of income for a bank. A term deposit is a cash investment held at a financial institution. Your money is invested for an agreed rate of interest over a fixed amount of time, or term. The bank has various outreach plans to sell term deposits to their customers such as email marketing, advertisements, telephonic marketing, and digital marketing.

Telephonic marketing campaigns still remain one of the most effective way to reach out to people. However, they require huge investment as large call centers are hired to actually execute these campaigns. Hence, it is crucial to identify the customers most likely to convert beforehand so that they can be specifically targeted via call.

The data is related to direct marketing campaigns (phone calls) of a Portuguese banking institution. The classification goal is to predict if the client will subscribe to a term deposit (variable y).

Content
The data is related to the direct marketing campaigns of a Portuguese banking institution. The marketing campaigns were based on phone calls. Often, more than one contact to the same client was required, in order to access if the product (bank term deposit) would be ('yes') or not ('no') subscribed by the customer or not. The data folder contains two datasets:-

Bank.csv: 45,211 rows and 18 columns ordered by date (from May 2008 to November 2010)

Detailed Column Descriptions
bank client data:

1 - age (numeric)

2 - job : type of job (categorical: "admin.","unknown","unemployed","management","housemaid","entrepreneur","student",
"blue-collar","self-employed","retired","technician","services")

3 - marital : marital status (categorical: "married","divorced","single"; note: "divorced" means divorced or widowed)

4 - education (categorical: "unknown","secondary","primary","tertiary")

5 - default: has credit in default? (binary: "yes","no")

6 - balance: average yearly balance, in euros (numeric)

7 - housing: has housing loan? (binary: "yes","no")

8 - loan: has personal loan? (binary: "yes","no")
# related with the last contact of the current campaign:
9 - contact: contact communication type (categorical: "unknown","telephone","cellular")
10 - day: last contact day of the month (numeric)

11 - month: last contact month of year (categorical: "jan", "feb", "mar", …, "nov", "dec")

12 - duration: last contact duration, in seconds (numeric)

# other attributes:
13 - campaign: number of contacts performed during this campaign and for this client (numeric, includes last contact)

14 - pdays: number of days that passed by after the client was last contacted from a previous campaign (numeric, -1 means client was not previously contacted)

15 - previous: number of contacts performed before this campaign and for this client (numeric)

16 - poutcome: outcome of the previous marketing campaign (categorical: "unknown","other","failure","success")

Output variable (desired target):

17 - y - has the client subscribed a term deposit? (binary: "yes","no")

Missing Attribute Values: None


In [3]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
df=pd.read_csv('https://raw.githubusercontent.com/ogut77/DataScience/main/data/Bank.csv',sep = ';')
df

age           job   marital  education default  balance housing loan  \
0       58    management   married   tertiary      no     2143     yes   no   
1       44    technician    single  secondary      no       29     yes   no   
2       33  entrepreneur   married  secondary      no        2     yes  yes   
3       47   blue-collar   married    unknown      no     1506     yes   no   
4       33       unknown    single    unknown      no        1      no   no   
...    ...           ...       ...        ...     ...      ...     ...  ...   
45206   51    technician   married   tertiary      no      825      no   no   
45207   71       retired  divorced    primary      no     1729      no   no   
45208   72       retired   married  secondary      no     5715      no   no   
45209   57   blue-collar   married  secondary      no      668      no   no   
45210   37  entrepreneur   married  secondary      no     2971      no   no   

         contact  day month  duration  campaign  pdays  previous poutcome    y  
0        unknown    5   may       261         1     -1         0  unknown   no  
1        unknown    5   may       151         1     -1         0  unknown   no  
2        unknown    5   may        76         1     -1         0  unknown   no  
3        unknown    5   may        92         1     -1         0  unknown   no  
4        unknown    5   may       198         1     -1         0  unknown   no  
...          ...  ...   ...       ...       ...    ...       ...      ...  ...  
45206   cellular   17   nov       977         3     -1         0  unknown  yes  
45207   cellular   17   nov       456         2     -1         0  unknown  yes  
45208   cellular   17   nov      1127         5    184         3  success  yes  
45209  telephone   17   nov       508         4     -1         0  unknown   no  
45210   cellular   17   nov       361         2    188        11    other   no  

[45211 rows x 17 columns]

In [4]:
print(df.shape)
df.info()
df.isnull().sum()

(45211, 17)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45211 entries, 0 to 45210
Data columns (total 17 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   age        45211 non-null  int64 
 1   job        45211 non-null  object
 2   marital    45211 non-null  object
 3   education  45211 non-null  object
 4   default    45211 non-null  object
 5   balance    45211 non-null  int64 
 6   housing    45211 non-null  object
 7   loan       45211 non-null  object
 8   contact    45211 non-null  object
 9   day        45211 non-null  int64 
 10  month      45211 non-null  object
 11  duration   45211 non-null  int64 
 12  campaign   45211 non-null  int64 
 13  pdays      45211 non-null  int64 
 14  previous   45211 non-null  int64 
 15  poutcome   45211 non-null  object
 16  y          45211 non-null  object
dtypes: int64(7), object(10)
memory usage: 5.9+ MB


age          0
job          0
marital      0
education    0
default      0
balance      0
housing      0
loan         0
contact      0
day          0
month        0
duration     0
campaign     0
pdays        0
previous     0
poutcome     0
y            0
dtype: int64

In [5]:
#For object check the data 
for cn in df.columns:
  if(df[cn].dtype==object):
    print(df[cn].value_counts())
  

blue-collar      9732
management       9458
technician       7597
admin.           5171
services         4154
retired          2264
self-employed    1579
entrepreneur     1487
unemployed       1303
housemaid        1240
student           938
unknown           288
Name: job, dtype: int64
married     27214
single      12790
divorced     5207
Name: marital, dtype: int64
secondary    23202
tertiary     13301
primary       6851
unknown       1857
Name: education, dtype: int64
no     44396
yes      815
Name: default, dtype: int64
yes    25130
no     20081
Name: housing, dtype: int64
no     37967
yes     7244
Name: loan, dtype: int64
cellular     29285
unknown      13020
telephone     2906
Name: contact, dtype: int64
may    13766
jul     6895
aug     6247
jun     5341
nov     3970
apr     2932
feb     2649
jan     1403
oct      738
sep      579
mar      477
dec      214
Name: month, dtype: int64
unknown    36959
failure     4901
other       1840
success     1511
Name: poutcome, dtype: int64
n

In [6]:
def Encoder(df):
          from sklearn import preprocessing
          columnsToEncode = list(df.select_dtypes(include=['category','object']))
          le = preprocessing.LabelEncoder()
          for feature in columnsToEncode:
              try:
                  df[feature] = le.fit_transform(df[feature])
              except:
                  print('Error encoding '+feature)
          return df


In [7]:
df=Encoder(df)
df

age  job  marital  education  default  balance  housing  loan  contact  \
0       58    4        1          2        0     2143        1     0        2   
1       44    9        2          1        0       29        1     0        2   
2       33    2        1          1        0        2        1     1        2   
3       47    1        1          3        0     1506        1     0        2   
4       33   11        2          3        0        1        0     0        2   
...    ...  ...      ...        ...      ...      ...      ...   ...      ...   
45206   51    9        1          2        0      825        0     0        0   
45207   71    5        0          0        0     1729        0     0        0   
45208   72    5        1          1        0     5715        0     0        0   
45209   57    1        1          1        0      668        0     0        1   
45210   37    2        1          1        0     2971        0     0        0   

       day  month  duration  campaign  pdays  previous  poutcome  y  
0        5      8       261         1     -1         0         3  0  
1        5      8       151         1     -1         0         3  0  
2        5      8        76         1     -1         0         3  0  
3        5      8        92         1     -1         0         3  0  
4        5      8       198         1     -1         0         3  0  
...    ...    ...       ...       ...    ...       ...       ... ..  
45206   17      9       977         3     -1         0         3  1  
45207   17      9       456         2     -1         0         3  1  
45208   17      9      1127         5    184         3         2  1  
45209   17      9       508         4     -1         0         3  0  
45210   17      9       361         2    188        11         1  0  

[45211 rows x 17 columns]

In [8]:
y = df['y'] #Output
X = df.drop('y',axis=1)
X

age  job  marital  education  default  balance  housing  loan  contact  \
0       58    4        1          2        0     2143        1     0        2   
1       44    9        2          1        0       29        1     0        2   
2       33    2        1          1        0        2        1     1        2   
3       47    1        1          3        0     1506        1     0        2   
4       33   11        2          3        0        1        0     0        2   
...    ...  ...      ...        ...      ...      ...      ...   ...      ...   
45206   51    9        1          2        0      825        0     0        0   
45207   71    5        0          0        0     1729        0     0        0   
45208   72    5        1          1        0     5715        0     0        0   
45209   57    1        1          1        0      668        0     0        1   
45210   37    2        1          1        0     2971        0     0        0   

       day  month  duration  campaign  pdays  previous  poutcome  
0        5      8       261         1     -1         0         3  
1        5      8       151         1     -1         0         3  
2        5      8        76         1     -1         0         3  
3        5      8        92         1     -1         0         3  
4        5      8       198         1     -1         0         3  
...    ...    ...       ...       ...    ...       ...       ...  
45206   17      9       977         3     -1         0         3  
45207   17      9       456         2     -1         0         3  
45208   17      9      1127         5    184         3         2  
45209   17      9       508         4     -1         0         3  
45210   17      9       361         2    188        11         1  

[45211 rows x 16 columns]

In [9]:
from sklearn.model_selection import  train_test_split


X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=17)

 


Q1)Using  Random Forest,XGBoost, Light GBM and Gradient Boosting Classifier with default parameters (no parameter specifications except random_state) calculate Accuracy on Test data. Which method gives the best accuracy on test data

In [10]:
from sklearn.metrics import accuracy_score
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import GradientBoostingClassifier
from time import time
from xgboost import XGBClassifier
import lightgbm as lgb

rfc = RandomForestClassifier(criterion = 'entropy', random_state = 17)
rfc.fit(X_train, y_train)
rfc_pred_test = rfc.predict(X_test)
print('Testing Set Evaluation Accuracy score Random Forest=>',accuracy_score(y_test,rfc_pred_test))

lgb_model = lgb.LGBMClassifier(random_state=17)
lgb_model.fit(X_train, y_train)
print('Testing Set Evaluation Accuracy score Light GBM=>', accuracy_score(y_test, lgb_model.predict(X_test)))

model = XGBClassifier(random_state=17)
model.fit(X_train, y_train)
xgb_pred = model.predict(X_test)
print('Testing Set Evaluation Accuracy score XGBoost=>',accuracy_score(y_test, xgb_pred))

gbm_model = GradientBoostingClassifier(random_state=17)
gbm_model.fit(X_train, y_train)
print('Testing Set Evaluation Accuracy score Gradient Boosting=>',accuracy_score(y_test, gbm_model.predict(X_test)))

Testing Set Evaluation Accuracy score Random Forest=> 0.9031230646730957
Testing Set Evaluation Accuracy score Light GBM=> 0.9073697248518092
Testing Set Evaluation Accuracy score XGBoost=> 0.9043616738918872
Testing Set Evaluation Accuracy score Gradient Boosting=> 0.9018844554543042


ANSWER: *The best accuracy score was from Light GBM model, then XGBoost, Random Forest, and the least accurate was from Gradient Boosting.*

Q2) Using optuna hyperparmeter optimization technique and 100 trial

 a)find best methods with  parameters  using Cross validation (CV=3) technique for the range of   parameters below. What are the best parameters for the method with highest cross validation accuracy?

"max_depth": range(2, 16), "max_features": range(2, 16)

 b)Evaluate the performance of the  method with highest cross validation accuracy on test data.What is the accuracy value?


In [16]:
from sklearn.model_selection import GridSearchCV, cross_val_score

rfc_params = {"max_depth": range(2, 16), "max_features": range(2, 16)}
rfc_grid = GridSearchCV(rfc, rfc_params, cv=3, n_jobs=100, verbose=True)
rfc_grid.fit(X_train, y_train)
rfc_grid.best_params_ 
accuracy_score(y_test, rfc_grid.predict(X_test))

Fitting 3 folds for each of 196 candidates, totalling 588 fits


0.9047155622401133

For Q3 and Q4 ,use the following data.

In [10]:
dr=pd.read_csv('https://raw.githubusercontent.com/ogut77/DataScience/main/data/diamond.csv')
dr

Carat Weight              Cut Color Clarity Polish Symmetry Report  \
0             1.10            Ideal     H     SI1     VG       EX    GIA   
1             0.83            Ideal     H     VS1     ID       ID   AGSL   
2             0.85            Ideal     H     SI1     EX       EX    GIA   
3             0.91            Ideal     E     SI1     VG       VG    GIA   
4             0.83            Ideal     G     SI1     EX       EX    GIA   
...            ...              ...   ...     ...    ...      ...    ...   
5995          1.03            Ideal     D     SI1     EX       EX    GIA   
5996          1.00        Very Good     D     SI1     VG       VG    GIA   
5997          1.02            Ideal     D     SI1     EX       EX    GIA   
5998          1.27  Signature-Ideal     G     VS1     EX       EX    GIA   
5999          2.19            Ideal     E     VS1     EX       EX    GIA   

      Price  
0      5169  
1      3470  
2      3183  
3      4370  
4      3171  
...     ...  
5995   6250  
5996   5328  
5997   6157  
5998  11206  
5999  30507  

[6000 rows x 8 columns]

In [11]:
def Encoder(df):
          from sklearn import preprocessing
          columnsToEncode = list(df.select_dtypes(include=['category','object']))
          le = preprocessing.LabelEncoder()
          for feature in columnsToEncode:
              try:
                  df[feature] = le.fit_transform(df[feature])
              except:
                  print('Error encoding '+feature)
          return df


In [12]:
dr=Encoder(dr)
dr

Carat Weight  Cut  Color  Clarity  Polish  Symmetry  Report  Price
0             1.10    2      4        2       3         0       1   5169
1             0.83    2      4        3       2         2       0   3470
2             0.85    2      4        2       0         0       1   3183
3             0.91    2      1        2       3         3       1   4370
4             0.83    2      3        2       0         0       1   3171
...            ...  ...    ...      ...     ...       ...     ...    ...
5995          1.03    2      0        2       0         0       1   6250
5996          1.00    4      0        2       3         3       1   5328
5997          1.02    2      0        2       0         0       1   6157
5998          1.27    3      3        3       0         0       1  11206
5999          2.19    2      1        3       0         0       1  30507

[6000 rows x 8 columns]

In [13]:
y = dr['Price'] #Output
X = dr.drop('Price',axis=1)
X

Carat Weight  Cut  Color  Clarity  Polish  Symmetry  Report
0             1.10    2      4        2       3         0       1
1             0.83    2      4        3       2         2       0
2             0.85    2      4        2       0         0       1
3             0.91    2      1        2       3         3       1
4             0.83    2      3        2       0         0       1
...            ...  ...    ...      ...     ...       ...     ...
5995          1.03    2      0        2       0         0       1
5996          1.00    4      0        2       3         3       1
5997          1.02    2      0        2       0         0       1
5998          1.27    3      3        3       0         0       1
5999          2.19    2      1        3       0         0       1

[6000 rows x 7 columns]

In [14]:
from sklearn.model_selection import  train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=17)

Q3)Using Linear Regression,Decison Tree Random Forest,XGBoost, Light GBM and Gradient Boosting Classifier with default parameters (no parameter specifications except random_state) calculate R2 statistics on test data. Which method gives the best accuracy on test data

In [22]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor
from xgboost import XGBRegressor
from lightgbm import LGBMRegressor
from sklearn.metrics import r2_score

models = {
    'Linear Regression': LinearRegression(),
    'Decision Tree': DecisionTreeRegressor(random_state=17),
    'Random Forest': RandomForestRegressor(random_state=17),
    'XGBoost': XGBRegressor(random_state=17),
    'Light GBM': LGBMRegressor(random_state=17),
    'Gradient Boosting': GradientBoostingRegressor(random_state=17)
}

for name, model in models.items():
    model.fit(X_train, y_train)
    y_pred = model.predict(X_test)
    r2 = r2_score(y_test, y_pred)
    print(f'{name}: R2 score = {r2:.4f}')

Linear Regression: R2 score = 0.8234
Decision Tree: R2 score = 0.9592
Random Forest: R2 score = 0.9800
XGBoost: R2 score = 0.9803
Light GBM: R2 score = 0.9813
Gradient Boosting: R2 score = 0.9739


ANSWER: *We can see that all six models perform well on the test set, with Light GBM having the highest R2 score of 0.9813.*

Q4) Using optuna hyperparmeter optimization technique (100 trial)  with Random Forest,XGBoost, Light GBM and Gradient Boosting Regressor

a)find best methods with  parameters  using Cross validation (CV=3) technique for the range of   parameters below. What are the best parameters for the method with highest cross validation R2?

"max_depth": range(2, 7), "max_features": range(2, 7)

 b)Evaluate the performance of the  method with highest cross validation R2 on test data. What is the R2 value?


In [17]:
!pip install optuna

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 365.7/365.7 KB 7.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 212.3/212.3 KB 23.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 78.7/78.7 KB 11.5 MB/s eta 0:00:00


In [ ]:
import pandas as pd
from sklearn.model_selection import cross_val_score
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor
import xgboost as xgb
import lightgbm as lgb
import optuna

def objective(trial):
    
    method = trial.suggest_categorical('method', ['RandomForest', 'GradientBoosting', 'XGBoost', 'LightGBM'])

    if method == 'RandomForest':
        params = {
            'max_depth': trial.suggest_int('max_depth', 2, 7),
            'max_features': trial.suggest_int('max_features', 2, 7),
            'random_state': 17
        }
        model = RandomForestRegressor(**params)

    elif method == 'GradientBoosting':
        params = {
            'max_depth': trial.suggest_int('max_depth', 2, 7),
            'max_features': trial.suggest_int('max_features', 2, 7),
            'random_state': 17
        }
        model = GradientBoostingRegressor(**params)

    elif method == 'XGBoost':
        params = {
            'max_depth': trial.suggest_int('max_depth', 2, 7),
            'max_features': trial.suggest_int('max_features', 2, 7),
            'random_state': 17,
            'objective': 'reg:squarederror'
        }
        model = xgb.XGBRegressor(**params)

    elif method == 'LightGBM':
        params = {
            'max_depth': trial.suggest_int('max_depth', 2, 7),
            'max_features': trial.suggest_int('max_features', 2, 7),
            'random_state': 17
        }
        model = lgb.LGBMRegressor(**params)

    cv_score = cross_val_score(model, diamonds.drop('price', axis=1), diamonds['price'], cv=3, scoring='r2').mean()

    return -cv_score

study = optuna.create_study(direction='minimize')
study.optimize(objective, n_trials=100)

best_params = study.best_params
best_score = -study.best_value

print(f'Best method: {study.best_trial.user_attrs["best_method"]}')
print(f'Best parameters: {best_params}')
print(f'Best cross-validation R2 score: {best_score:.3f}')

In [ ]:
from sklearn.model_selection import train_test_split
train, test = train_test_split(dr, test_size=0.2, random_state=17)

if study.best_trial.user_attrs['best_method'] == 'RandomForest':
    model = RandomForestRegressor(**best_params)
elif study.best_trial.user_attrs['best_method']